streams/common/src/main/java/com/hortonworks/streamline/streams/common/StreamlineEventImpl.java

```diff
@@ -198,34 +175,20 @@ public StreamlineEventImpl build() {
         return new StreamlineEventImpl.Builder();
     }
 
-    public StreamlineEventImpl(StreamlineEventImpl other) {
-        this.header = other.header;
-        this.sourceStream = other.sourceStream;
-        this.auxiliaryFieldsAndValues = new HashMap<>(other.auxiliaryFieldsAndValues);
-        this.dataSourceId = other.dataSourceId;
-        this.id = other.id;
-        this.delegate = ImmutableMap.copyOf(other.delegate);
-    }
-
-    /*
-     * Creates a copy of 'other' but with the given keyValues.
-     */
-    private StreamlineEventImpl(StreamlineEventImpl other, ImmutableMap<String, Object> keyValues) {
-        this.header = other.header;
-        this.sourceStream = other.sourceStream;
-        this.auxiliaryFieldsAndValues = new HashMap<>(other.auxiliaryFieldsAndValues);
-        this.dataSourceId = other.dataSourceId;
-        this.id = other.id;
-        this.delegate = ImmutableMap.copyOf(keyValues);
-    }
-
     @Override
     public Map<String, Object> getAuxiliaryFieldsAndValues() {
         return auxiliaryFieldsAndValues;
     }
 
-    public void addAuxiliaryFieldAndValue(String field, Object value) {
-        auxiliaryFieldsAndValues.put(field, value);
+    @Override
+    public StreamlineEvent addAuxiliaryFieldAndValue(String field, Object value) {
+        Map<String, Object> aux = new HashMap<>();
```

Comment:
Use an immutable map here so the build below can just use it and avoid the extra HashMap?

Reply:
Actually I went a same path, but changed the approach cause ImmutableMap builder doesn't allow putting same key more than once. It threw exception at runtime.

Reply:
FYI copying part of source code in ImmutableMap.java

```
    /**
     * Adds the given {@code entry} to the map, making it immutable if
     * necessary. Duplicate keys are not allowed, and will cause {@link #build}
     * to fail.
     *
     * @since 11.0
     */
    public Builder<K, V> put(Entry<? extends K, ? extends V> entry) {
      return put(entry.getKey(), entry.getValue());
    }

    /**
     * Associates all of the given map's keys and values in the built map.
     * Duplicate keys are not allowed, and will cause {@link #build} to fail.
     *
     * @throws NullPointerException if any key or value in {@code map} is null
     */
    public Builder<K, V> putAll(Map<? extends K, ? extends V> map) {
      ensureCapacity(size + map.size());
      for (Entry<? extends K, ? extends V> entry : map.entrySet()) {
        put(entry);
      }
      return this;
    }
```

Worth noting that the map doesn't allow null keys as well as **null values**, as javadoc of class stated.